### PR TITLE
Message Reply UI test

### PR DESCRIPTION
**Pull request by shelbaz**
Issue #147

### CODE REVIEW

File: test/androidTest/java/org/thoughtcrime/securesms/RepliesEspressoTest.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    public ActivityTestRule<ConversationListActivity> mainActivityRule =
+            new ActivityTestRule(ConversationListActivity.class, true, false);
+
+    @Test
+    public void buttonExists(){
+        Helper helper = new Helper(mainActivityRule);
+        String testSend = helper.randString();
+
+        helper
+                .goConversations()
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/RepliesEspressoTest.java]: please de-indent all page navigation methods `go*`

File: test/androidTest/java/org/thoughtcrime/securesms/RepliesEspressoTest.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+@RunWith(AndroidJUnit4.class)
+@LargeTest
+public class RepliesEspressoTest {
+    @Rule
+    public ActivityTestRule<ConversationListActivity> mainActivityRule =
+            new ActivityTestRule(ConversationListActivity.class, true, false);
+
+    @Test
+    public void buttonExists(){
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/RepliesEspressoTest.java]: `() {`

File: test/androidTest/java/org/thoughtcrime/securesms/RepliesEspressoTest.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        Helper helper = new Helper(mainActivityRule);
+        String testSend = helper.randString();
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/RepliesEspressoTest.java]: The two last tests have a space between the helper and the test string(s). For consistency, I think they should all be the same (an I prefer having the space)

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java [hunk 4]
```diff
@@ -89,6 +93,36 @@ public ConversationHelper sendReaction(int position, int emoji) {
         return this;
     }
 
+    public ConversationHelper sendReply(int position, String message) {
+
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java]: please remove extra line

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java [hunk 5]
```diff
@@ -89,6 +93,36 @@ public ConversationHelper sendReaction(int position, int emoji) {
...
+        this.unselectMessage();
+        this.selectMessage(position);
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java]: Following the pattern in the rest of the file, could you add a line between these two method calls

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java [hunk 6]
```diff
@@ -89,6 +93,36 @@ public ConversationHelper sendReaction(int position, int emoji) {
...
+                .perform(android.support.test.espresso.action.ViewActions.pressKey(KeyEvent.KEYCODE_BACK));
+
+        clearKeyboard();
+
+        return this;
+    }
+
+    public ConversationHelper clearKeyboard() {
+        onView(withId(R.id.embedded_text_editor))
+                .perform(click())
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java]: please de-indent these two calls

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java [hunk 7]
```diff
@@ -89,6 +93,36 @@ public ConversationHelper sendReaction(int position, int emoji) {
...
+        onView(withId(R.id.custom_reply))
+                .perform(typeText(message))
+
+                .perform(android.support.test.espresso.action.ViewActions.pressKey(KeyEvent.KEYCODE_ENTER))
+                .perform(replaceText(""))
+                .perform(android.support.test.espresso.action.ViewActions.pressKey(KeyEvent.KEYCODE_BACK))
+                .perform(android.support.test.espresso.action.ViewActions.pressKey(KeyEvent.KEYCODE_BACK));
+
+        clearKeyboard();
+
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java]: Don't forget to reset the message selected state

```java
this.messageSelected = false;
```

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java [hunk 8]
```diff
@@ -89,6 +93,36 @@ public ConversationHelper sendReaction(int position, int emoji) {
         return this;
     }
 
+    public ConversationHelper sendReply(int position, String message) {
+
+        this.unselectMessage();
+        this.selectMessage(position);
+
+        onView(withId(R.id.menu_context_reply))
+                .perform(click());
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java]: please de-indent all of these perform calls

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java [hunk 9]
```diff
@@ -89,6 +93,36 @@ public ConversationHelper sendReaction(int position, int emoji) {
...
+        onView(withId(R.id.lin_reply))
+                .perform(click());
+        onView(withId(R.id.custom_reply))
+                .perform(typeText(message))
+
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java]: is this space intentional? it feels awkward

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java [hunk 10]
```diff
@@ -89,6 +93,36 @@ public ConversationHelper sendReaction(int position, int emoji) {
...
+                .perform(android.support.test.espresso.action.ViewActions.pressKey(KeyEvent.KEYCODE_ENTER))
+                .perform(replaceText(""))
+                .perform(android.support.test.espresso.action.ViewActions.pressKey(KeyEvent.KEYCODE_BACK))
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java]: Could you explain what these `KEYCODE_BACK` events are doing?

**Review by g-harel — CHANGES_REQUESTED** [review]

File: test/androidTest/java/org/thoughtcrime/securesms/RepliesEspressoTest.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+        String testSend = helper.randString();
+        String testReply = helper.randString();
+
+        helper
+                .goConversations()
+                .goConversation()
+                .sendMessage(testSend)
+                .sendReply(0, testReply)
+                .assertText(testReply);
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/RepliesEspressoTest.java]: this asserts that the reply is visible, but not necessarily that it is in the correct message. Have you tried using `assertTextAt`?

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java (hunk 9 above)
Comment (shelbaz) [reply test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java]: yes its for readability

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java (hunk 10 above)
Comment (shelbaz) [reply test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java]: 1st keycode back exits the keyboard, 2nd exits reply mode. This is the only way to exit reply mode currently that i know of.

File: test/androidTest/java/org/thoughtcrime/securesms/RepliesEspressoTest.java (hunk 11 above)
Comment (shelbaz) [reply test/androidTest/java/org/thoughtcrime/securesms/RepliesEspressoTest.java]: Tried, had a lot of trouble with it. Settled with this , and generated random strings instead to skirt around this. Assumed if its displaying, its in the right spot. Naive, maybe.

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java [hunk 12]
```diff
@@ -89,6 +93,36 @@ public ConversationHelper sendReaction(int position, int emoji) {
...
+
+        this.selectMessage(position);
+
+        onView(withId(R.id.menu_context_reply))
+            .perform(click());
+        onView(withId(R.id.lin_reply))
+            .perform(click());
+        onView(withId(R.id.custom_reply))
+            .perform(typeText(message))
+
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java:107]: Is this space intentional?

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java [hunk 13]
```diff
@@ -89,6 +93,36 @@ public ConversationHelper sendReaction(int position, int emoji) {
...
+            .perform(android.support.test.espresso.action.ViewActions.pressKey(KeyEvent.KEYCODE_ENTER))
+            .perform(replaceText(""))
+            .perform(android.support.test.espresso.action.ViewActions.pressKey(KeyEvent.KEYCODE_BACK))
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java:110]: What are these last two lines doing?

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java (hunk 12 above)
Comment (shelbaz) [reply test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java:107]: yes, readability

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java (hunk 13 above)
Comment (shelbaz) [reply test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java:110]: 1st back hides keyboard, 2nd back exits reply mode. Could not find a way to exit search mode besides that

**Review by g-harel — APPROVED** [review]

**Pull request merged**